src/test/socket/listen/test_listen.rs

```diff
@@ -0,0 +1,434 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+struct ListenArguments {
+    fd: libc::c_int,
+    backlog: libc::c_int,
+}
+
+#[derive(Debug, Copy, Clone)]
+struct BindAddress {
+    address: libc::in_addr_t,
+    port: libc::in_port_t,
+}
+
+/// A boxed function to run as a test.
+type TestFn = Box<dyn Fn() -> Result<(), String>>;
+
+fn main() {
+    // should we run only tests that shadow supports
+    let run_only_passing_tests = match std::env::args().position(|x| x == "--shadow-passing") {
+        Some(_) => true,
+        None => false,
+    };
+    // should we summarize the results rather than exit on a failed test
+    let summarize = match std::env::args().position(|x| x == "--summarize") {
+        Some(_) => true,
+        None => false,
+    };
+
+    let tests = if run_only_passing_tests {
+        get_passing_tests()
+    } else {
+        get_all_tests()
+    };
+
+    if let Err(_) = run_tests(tests.iter(), summarize) {
```

Comment:
I haven't tried it, but I saw someone mentioning in passing that `main` can return a Result, in which case you just return the result from this call. (Not sure what the output would look like on error though; explicit handling like you have it might still be better)

Reply:
I tried this again and the output doesn't look too bad, so I made this change.